Enable portable mode if "profile" folder exists

Tester798 opened the pull request:
Enable portable mode not only by adding command line option but also by checking if "profile" folder exists in the app dir. Closes #9445.

sledgehammer999 commented on src/app/cmdoptions.cpp:
```diff
@@ -36,6 +36,8 @@
 #include <QProcessEnvironment>
 #include <QTextStream>
+#include <QCoreApplication>
+#include <QDir>
```
Add them in alphabetical order

sledgehammer999 commented on src/app/cmdoptions.cpp:
```diff
@@ -481,6 +484,10 @@ QBtCommandLineParameters parseCommandLine(const QStringList &args)
         }
     }
 
+    if (QDir(QCoreApplication::applicationDirPath()).exists(DEFAULT_PORTABLE_MODE_PROFILE_DIR)) {
```
Curly braces aren't needed in this if statement.
https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md#e-single-statement-if-blocks

Chocobo1 commented on src/app/cmdoptions.cpp:
```diff
@@ -54,6 +56,7 @@ namespace
     const int USAGE_INDENTATION = 4;
     const int USAGE_TEXT_COLUMN = 31;
     const int WRAP_AT_COLUMN = 80;
+    const QString DEFAULT_PORTABLE_MODE_PROFILE_DIR = QStringLiteral("profile");
```
There is src/base/profile.cpp file so this declaration should not be here.

Tester798 replied:
Added this because there is same piece here: https://github.com/qbittorrent/qBittorrent/blob/master/src/app/application.cpp#L118

Should I try to move it all to src/base/profile.cpp ?

Chocobo1 replied:
>Should I try to move it all to src/base/profile.cpp ?

Maybe yes, at least it should try to fit in the helper functions that file provides. Let us wait for others opinion.

glassez commented on src/app/cmdoptions.cpp:
```diff
@@ -481,6 +484,9 @@ QBtCommandLineParameters parseCommandLine(const QStringList &args)
         }
     }
 
+    if (QDir(QCoreApplication::applicationDirPath()).exists(DEFAULT_PORTABLE_MODE_PROFILE_DIR))
```
IMO, it's not "Command line options" job.

Kolcha replied:
Agreed, checking some folder for existence is not related to command line parsing. I suggest to do that after command line parsing

Tester798 replied:
Do you mean it's better to add folder check here: https://github.com/qbittorrent/qBittorrent/blob/565eb4cdc8631ce0e8a999bd1234d5ce67baf990/src/app/application.cpp#L149 ?

Should I also add folder checks here:
https://github.com/qbittorrent/qBittorrent/blob/565eb4cdc8631ce0e8a999bd1234d5ce67baf990/src/app/application.cpp#L743 and here:
https://github.com/qbittorrent/qBittorrent/blob/565eb4cdc8631ce0e8a999bd1234d5ce67baf990/src/app/application.cpp#L746 or these two places are ok to check only for `m_commandLineArgs.portableMode`?

Kolcha replied:
just do it in only one place, right exactly before the line you posted first. just set `m_commandLineArgs.portableMode` to `true` in case if folder exists, and leave everything as is. it must work

glassez replied:
>just set m_commandLineArgs.portableMode to true

Spoofing command line parameters doesn't seem like the better solution. Ideally, it should be checked in the place where the final decision is made.

Tester798 replied:
I think it's better to set `m_commandLineArgs.portableMode` to `true` before https://github.com/Tester798/qBittorrent/blob/e9e30993eebbc7f678af3632a604a7584b2df598/src/app/application.cpp#L150 because in this case we can still get useful messages from https://github.com/Tester798/qBittorrent/blob/e9e30993eebbc7f678af3632a604a7584b2df598/src/app/application.cpp#L746-L750
Updated pull request.

Tester798 commented:
Ok, here is the variant without "Spoofing command line parameters" way.

glassez commented:
It looks like this feature makes `--portable` command line parameter completely meaningless, so it should be removed. Or am I missing some important use case?

Tester798 commented:
Hmm.. Yes, after first time when program will be run with `--portable` command line parameter `profile` folder will be created in app folder and use of `--portable` command line parameter wouldn't be needed anymore.
But if `--portable` command line parameter will be removed, then user will need to create `profile` folder himself manually to initiate portable operation, so I'm not sure if the parameter should be removed.

The intent of this pull request is just to add the possibility to run the program in portable mode without need of specifying any additional portable parameters.

glassez commented:
>But if --portable command line parameter will be removed, then user will need to create profile folder himself manually to initiate portable operation, so I'm not sure if the parameter should be removed.

Creating folder is no harder than running program with some parameter.

Tester798 commented:
So should I add removing --portable command line parameter to this pull request too?

glassez commented:
>So should I add removing --portable command line parameter to this pull request too?

Yes. Unless someone provides a compelling reason not to. Let's wait at least for @Chocobo1 opinion.

Chocobo1 commented:
>Yes. Unless someone provides a compelling reason not to. Let's wait at least for @Chocobo1 opinion.

I don't have opinion on either way under the condition that `--profile` switch won't be affected, otherwise we broke some usage pattern here.

Also I would like to see a new log entry stating the used profile directory when qbt is started (whether in portable mode or not, generally useful for users IMO), can you do that @Tester798?

Chocobo1 commented on src/app/application.cpp:
```diff
@@ -146,11 +146,11 @@ Application::Application(const QString &id, int &argc, char **argv)
         : m_commandLineArgs.profileDir;
     Profile::initialize(profileDir, m_commandLineArgs.configurationName,
-                        m_commandLineArgs.relativeFastresumePaths || m_commandLineArgs.portableMode);
+                        m_commandLineArgs.relativeFastresumePaths || isPortableModeEnabled());
```
you are checking the profile folder every time you call this function! you should only do that once.

Tester798 replied:
Replaced function call with `m_portableModeEnabled`:
https://github.com/Tester798/qBittorrent/blob/46b028841ebdb7f4e613f68b2b59a16cc8c0c095/src/app/application.cpp#L147

glassez commented:
>I don't have opinion on either way under the condition that --profile switch won't be affected, otherwise we broke some usage pattern here.

I.e. portable mode with profile dir other than default one is really useful? Do we allow such configuration currently? If No, I will insist on removing "--portable" param.

One important thing is @sledgehammer999 general approval of this way of "portable mode" detection. Several different options were discussed earlier (explicit portable build, params file, etc.) but I do not recall that we came to a consensus, because we still have open PRs that implement other ways.

Chocobo1 commented:
>I.e. portable mode with profile dir other than default one is really useful? Do we allow such configuration currently?

Yes.
I don't use portable/profile switch and I'm just saying that some users will be affected.

>If No, I will insist on removing "--portable" param.

Just to be clear, it seems the `--portable` switch is just a shortcut for `--profile qbt_binary_path/profile`.

glassez commented:
>Just to be clear, it seems the --portable switch is just a shortcut for --profile qbt_binary_path/profile.

Well, then it should definitely be removed here. All other "useful" cases are covered by other parameters.

Tester798 commented:
Removed `--portable` command line argument.
Added logging of config profile directory (because logging just base profile directory will require more code modification).

glassez commented:
>logging just base profile directory will require more code modification

Isn't "profileDir" suitable?

Tester798 commented:
`profileDir` is empty in case of not portable operation.

Tester798 commented on src/app/application.cpp:
```diff
@@ -740,9 +743,9 @@ void Application::cleanup()
 {
-    if (m_commandLineArgs.portableMode && !m_commandLineArgs.profileDir.isEmpty())
+    if (m_portableModeEnabled && !m_commandLineArgs.profileDir.isEmpty())
         throw CommandLineParameterError(tr("Portable mode and explicit profile directory options are mutually exclusive"));
```
Maybe we should remove this check completely? Because if there is profile folder in app dir and user wants to specify different profile with --profile= command line argument then app will not start at all.

Tester798 replied:
Removed.

Tester798 commented on src/app/application.cpp:
```diff
@@ -144,13 +144,15 @@ Application::Application(const QString &id, int &argc, char **argv)
     validateCommandLineParameters();
 
-    const QString profileDir = m_commandLineArgs.portableMode
+    const QString profileDir = m_portableModeEnabled && m_commandLineArgs.profileDir.isEmpty()
```
Also here we seems to need `&& m_commandLineArgs.profileDir.isEmpty()` because else `profile` folder will be used instead of `m_commandLineArgs.profileDir`.

Chocobo1 commented on src/app/application.cpp:
```diff
@@ -171,6 +173,7 @@ Application::Application(const QString &id, int &argc, char **argv)
         m_fileLogger = new FileLogger(fileLoggerPath(), isFileLoggerBackup(), fileLoggerMaxSize(), isFileLoggerDeleteOld(), fileLoggerAge(), static_cast<FileLogger::FileLogAgeType>(fileLoggerAgeType()));
 
     Logger::instance()->addMessage(tr("qBittorrent %1 started", "qBittorrent v3.2.0alpha started").arg(QBT_VERSION));
+    Logger::instance()->addMessage(tr("Using config directory '%1'").arg(Profile::instance().location(SpecialFolder::Config)));
```
nitpick, I would use "Using config directory: %1"

Tester798 replied:
Changed.

Chocobo1 commented on src/app/application.cpp:
```diff
@@ -144,13 +144,15 @@ Application::Application(const QString &id, int &argc, char **argv)
         : m_commandLineArgs.profileDir;
     Profile::initialize(profileDir, m_commandLineArgs.configurationName,
-                        m_commandLineArgs.relativeFastresumePaths || m_commandLineArgs.portableMode);
+                        m_commandLineArgs.relativeFastresumePaths || m_portableModeEnabled);
```
Please help us fix the coding style by adding parentheses to the last parameter.

glassez replied:
Seems like wrong logic here.
IIRC, if user specifies profile dir via command line it will used even if default profile dir is found. But if user doesn't specify "relative fastresume" param then it doesn't want it to use.
I would change workflow to the following:
1. Process profile related command-line parameters first. If no one is used then
2. Try to find default profile folder.

Tester798 replied:
Not sure if I understand what the logic should be with this `relativeFastresumePaths` command line parameter. As I understand `portable` mode should always use `relativeFastresumePaths`, but specifying `--profile=` command line parameter doesn't mean that `relativeFastresumePaths` should be used if it's not specified. Is that correct? If so, then maybe moving check `&& m_commandLineArgs.profileDir.isEmpty()` from https://github.com/Tester798/qBittorrent/blob/e707748f7a56999697e6314ba60efdbc6198289a/src/app/application.cpp#L151 to https://github.com/Tester798/qBittorrent/blob/e707748f7a56999697e6314ba60efdbc6198289a/src/app/application.cpp#L147 would do it?

Tester798 replied:
Ok, moved `m_commandLineArgs.profileDir.isEmpty()` check to the top.
Let me know if I should change anything else.

Chocobo1 commented on src/app/application.cpp:
```diff
@@ -740,9 +743,6 @@ void Application::cleanup()
 
 void Application::validateCommandLineParameters()
```
Not your code, please help us fix the coding style by adding `const` to function:
`void Application::validateCommandLineParameters() const`

Tester798 replied:
Done.

glassez commented on src/app/application.cpp:
```diff
@@ -144,13 +144,15 @@ Application::Application(const QString &id, int &argc, char **argv)
     QPixmapCache::setCacheLimit(PIXMAP_CACHE_SIZE);
 #endif
 
+    m_portableModeEnabled = QDir(QCoreApplication::applicationDirPath()).exists(DEFAULT_PORTABLE_MODE_PROFILE_DIR);
```
Please rename it to reflect its actual meaning: m_defaultProfileFolderFound. Current name confuses.

Tester798 replied:
Done.

glassez commented:
>Also "relative fastresume" option has no meaning without "profile" option

Sorry, it seems I'm wrong. It can have a meaning even with default profile (not portable).

So now your code seems to be correct, but `m_defaultProfileFolderFound` now has other meaning (default portable profile folder is found and isn't overriden by command-line parameters, so "portable mode enabled").

glassez commented on src/app/application.cpp:
```diff
@@ -144,13 +144,15 @@ Application::Application(const QString &id, int &argc, char **argv)
     QPixmapCache::setCacheLimit(PIXMAP_CACHE_SIZE);
 #endif
 
+    m_defaultProfileFolderFound = m_commandLineArgs.profileDir.isEmpty() && QDir(QCoreApplication::applicationDirPath()).exists(DEFAULT_PORTABLE_MODE_PROFILE_DIR);
```
>Also "relative fastresume" option has no meaning without "profile" option

Sorry, it seems I'm wrong. It can have a meaning even with default profile (not portable).

So now your code seems to be correct, but `m_defaultProfileFolderFound` now has other meaning (default portable profile folder is found and isn't overriden by command-line parameters, so "portable mode enabled").

Tester798 replied:
Thanks for the suggestions. I changed the code, please let me know if I should change anything else.

glassez commented on src/app/application.cpp:
```diff
@@ -171,6 +173,7 @@ Application::Application(const QString &id, int &argc, char **argv)
         m_fileLogger = new FileLogger(fileLoggerPath(), isFileLoggerBackup(), fileLoggerMaxSize(), isFileLoggerDeleteOld(), fileLoggerAge(), static_cast<FileLogger::FileLogAgeType>(fileLoggerAgeType()));
 
     Logger::instance()->addMessage(tr("qBittorrent %1 started", "qBittorrent v3.2.0alpha started").arg(QBT_VERSION));
+    Logger::instance()->addMessage(tr("Using config directory: %1").arg(Profile::instance().location(SpecialFolder::Config)));
```
I would add
```c++
if (portableModeEnabled)
    LogMsg(tr("Running in portable mode (auto detected)"));
```
Also I would move appropriate check from "validateCommandLineParameters" here (since it doesn't really validate command-line parameters themselves) so log messages will in correct order ("portable mode enabled" and then "relative fastresume is implied"). Also we will have no additional member variable.

Tester798 replied:
Moved check from "validateCommandLineParameters" here added additional log message. Please let me know if I should change anything else.

Chocobo1 commented on src/app/application.cpp:
```diff
@@ -171,6 +171,11 @@ Application::Application(const QString &id, int &argc, char **argv)
+    Logger::instance()->addMessage(tr("Using config directory: %1").arg(Profile::instance().location(SpecialFolder::Config)));
+    if (portableModeEnabled)
+        Logger::instance()->addMessage(tr("Running in portable mode (auto detected)"));
+    if (portableModeEnabled && m_commandLineArgs.relativeFastresumePaths)
```
Please make this inside `if (portableModeEnabled)`.
```c++
if (portableModeEnabled) {
    Logger::instance()->addMessage(tr("Running in portable mode (auto detected)"));
    if (m_commandLineArgs.relativeFastresumePaths)
        Logger::instance()->addMessage(tr("Portable mode implies relative fastresume"), Log::WARNING);
}
```

Chocobo1 commented on src/app/application.cpp:
```diff
@@ -171,6 +171,11 @@ Application::Application(const QString &id, int &argc, char **argv)
     Logger::instance()->addMessage(tr("qBittorrent %1 started", "qBittorrent v3.2.0alpha started").arg(QBT_VERSION));
+    Logger::instance()->addMessage(tr("Using config directory: %1").arg(Profile::instance().location(SpecialFolder::Config)));
+    if (portableModeEnabled)
+        Logger::instance()->addMessage(tr("Running in portable mode (auto detected)"));
```
will this work?
`tr("Running in portable mode. Auto detected profile folder at: %1").arg(profileDir)`

glassez replied:
It will show almost the same path as message above. Are you sure we need both of them?

Chocobo1 replied:
>Are you sure we need both of them?

I don't insist on that. But the point is `(auto detected)` has not enough information to users, they won't know what path is auto detected.

Chocobo1 replied:
Or maybe:
```c++
if (portableModeEnabled) {
    // ...
}
else {
    Logger::instance()->addMessage(tr("Using config directory: %1").arg(Profile::instance().location(SpecialFolder::Config)));
}
```

Chocobo1 commented on src/app/application.cpp:
```diff
@@ -171,6 +171,11 @@ Application::Application(const QString &id, int &argc, char **argv)
+    if (portableModeEnabled)
+        Logger::instance()->addMessage(tr("Running in portable mode (auto detected)"));
+    if (portableModeEnabled && m_commandLineArgs.relativeFastresumePaths)
+        Logger::instance()->addMessage(tr("Portable mode implies relative fastresume"), Log::WARNING);
```
```c++
tr("Redundant command line flag detected: \"%1\". Portable mode implies relative fastresume.").arg("--relative-fastresume")`  // to avoid translating the `--relative-fastresume` string
```

Chocobo1 commented:
@glassez 
Before we press the merge button. Should this land in v4.2.x or in next major release v4.3.0?

glassez commented:
I think it's not a problem to merge it in v4.2.x. v4.2.0 was not very successful and should not have spread much (I would treat it as one of the late beta/RC versions). Besides, hardly anyone has a "profile" folder in application directory for any other purpose than portable mode, so we are unlikely to break something.
Have you any other thoughts?

Chocobo1 commented:
>I think it's not a problem to merge it in v4.2.x. v4.2.0 was not very successful and should not have spread much (I would treat it as one of the late beta/RC versions). 

For me it seems many people have tried it and some (if not many) of them reverted back due to problems.

>Have you any other thoughts?

Not many, I just thought this is a behavior change and personally I would prefer including it in next major release instead (but I don't insist). Well I'll merge it and let @sledgehammer999 decide.

Chocobo1 commented:
@Tester798 
Thank you!

glassez commented:
>Well I'll merge it and let @sledgehammer999 decide.

:+1:

sledgehammer999 commented:
A lot of people were/interested for a painfree way to enable portable mode. So what's the worst case? That someone has forgotten a "profile" folder in the same folder as the binary? Big deal. Let's not forget that for official installations it is extremely hard for such a file to exist. (usually the binary is installed in protected/system folder).
So I am inclined to include this in the next release.

FreePlague commented:
How about QBT_PORTABLE = 1 environment variable? Is this gone too?

olliebean commented:
The first time I ran this new version without the --profile command line, all the downloaded files from my existing torrents - that had already been saved to the correct directories according to my Save Path in the Downloads tab of the options - were moved into the profile folder, and I had to move them back to where they should be by right-clicking on each one and selecting "Set location". Seems to have been a one-off, as new torrents are saving to the correct locations as per the settings, but surely this can't be working as designed?

glassez commented:
>The first time I ran this new version without the --profile command line,

It's not about --profile parameter. It's about --portable that is removed in favour of auto detection of "portable" mode.

olliebean commented:
I was using --profile=profile due to problems I had previously using the --portable switch. I just restarted qBittorrent without any command line, and it seems that problem still exists - all my torrents are now showing errors. The errors in the log read: "The filename, directory name, or volume label syntax is incorrect."

I guess I'll restore my configuration from backup and continue to use --profile=profile.

glassez commented:
--profile=profile doesn't mean "portable" mode. "portable" mode includes "relative fastresume paths".

olliebean commented:
Yes, and apparently relative fastresume paths doesn't work properly if those paths aren't in qBittorrent's directory. 4.2.1 means portable mode is automatically enabled if the "profile" folder is found in qBittorrent's home directory, so I'm going to have to either revert to an older version or rename the profile folder to get my paths working properly again.

olliebean commented:
Regardless, v4.2.1 seems to have royally messed up my existing torrents. It knows where they are - if I double-click on a torrent, it opens the correct directory with the correct file selected - but they are showing errors, and if I try a "recheck" it can't find them. Any guidance on how to get things working again would be welcomed. Restoring my profile folder from backup doesn't seem to have helped.

Janaue commented:
the issues for broken torrents is old https://github.com/qbittorrent/qBittorrent/issues/11178

olliebean commented:
Is there any way to explicitly disable relative fastresume when running portably? Previously the way to get around this was to use `--profile=profile` instead of --portable; now the workaround seems to be to use a different name for the profile directory, e.g., `--profile=userprofile`, to prevent `--relative-fastresume` being automatically enabled by the presence of the "profile" directory.

glassez commented:
When you set `--profile=profile` without `--relative-fastresume` it just will use profile folder you set.

*Note: I'm not sure qBittorrent can correctly handle existing torrents when you start it in another mode than the previous one.*
So your problems can be caused by switching between modes and then running qBittorrent with torrents created in another mode.

It seems that you do not understand exactly what has been done here. We haven't changed how it works in one mode or another. We just changed the way to enable "portable" mode.

olliebean commented:
Thanks - it seems my mistake was to run the executable directly, without using `--profile=profile`, upon which it corrupted the fastresume paths in my existing profile folder. I'd been using this command line option for so long, via a shortcut, that I'd forgotten that --portable mode has this problem (and if I'd remembered, I probably would have assumed it had been fixed by now anyway), and thought this new update meant I wouldn't need the command line option any more. Had I just kept using my shortcut as before, I wouldn't have run into any problems.

glassez commented:
@Tester798, can you confirm that "portable" mode works as expected? Do you use it currently?

Janaue commented:
> --portable

you do not need this flag now.  to enable portable mode you need a folder ''profile'' where the main exe is and it creates everything. Then there is still the bug about directory names.

Janaue commented:
so what is the proper way to enable the portable mode and avoid the error of directory name ?

olliebean commented:
The option `--profile=profile` is essentially the same as portable mode, but without the `--relative-fastresume` option that causes the path errors. It won't fix any path errors caused by running in portable mode, but it will prevent any more paths getting broken.

Tester798 commented:
@glassez didn't know it's released already, planned to switch to it after the this change.

Just checked versions 4.2.0 and 4.2.1 and looks like they work the same way in `portable` mode: giving error for the downloaded torrent after program is restarted if torrent was downloaded not to default folder.
If the program is running with `--profile=profile` command line, then looks like it's giving no error for the torrent after restarting the program (for both 4.2.0 and 4.2.1).

So looks like it's the problem with `--relative-fastresume` functionality?

Vanquo commented:
https://github.com/qbittorrent/qBittorrent/issues/7765
https://github.com/qbittorrent/qBittorrent/issues/11178
https://github.com/qbittorrent/qBittorrent/issues/11706
https://github.com/qbittorrent/qBittorrent/issues/7770
These are all somehow related. 

If I can describe the problem me, olliebean and others are having with portable mode, its trying to save torrents in the same drive letter as qbittorrent.exe but outside of the profile folder. Torrents saved in say ".\profile\qBittorrent\downloads" are fine after closing and reopening the app but any torrents located outside the profile folder but within the same drive letter will give the paths error. Any torrent saved in a different drive letter as qbittorrent.exe works fine.

Maybe portable mode should by default just not have `--relative-fastresume`. How about just keep the default to absolute paths and give us the option to decide if we want to save files in a relative path for portable use like specifying in the "Default Save Path" to save files in ".\profile\qBittorrent\Downloads" relative to the qbittorrent.exe.

sledgehammer999 commented:
Guys, this is a closed PR. If you experience problems, then open a new report describing the issue as best as you can. And maybe leave a link here too.

LeoKesler commented:
@Vanquo , the same happens if the torrents are saved in other driver letter. In this case, qbittorrent will copy the files to profile folder.

a-raccoon commented:
What is the explicit name of the "Profile Folder" in the program directory under Windows?  Assuming the following:

Drive = `E:`
ProgramDir = `E:\qBittorrent\`
AutoDetectPortableProfileDir = `E:\qBittorrent\????`

bacon-cheeseburger commented:
I know this is late in the conversation but there are some things people need to remember when it comes to software...

- User-defined paths should *always* be respected.
- No app should take it upon itself to move/copy files from user-defined paths. Trying to make apps "smart" by guessing what user intent is, almost always makes the app problematic. If I, as a user, want portable mode, *I* will tell the app, preferably by command line parameter. If I want to use a specific `profile`/dir for my config, *I* will tell you where to find it.
- K.I.S.S. *always* applies! Users who need/want to work outside of default settings know what they want. Stop trying to make assumptions, just listen to what they're telling you via the command line.
- GOOD apps do what the user asks, no more, no less. BAD apps argue with the user over what the user wants.

a-raccoon commented:
@bacon-cheeseburger : Not all users have command-lines or want to use command lines or shortcuts.  It is a very standard practice within the Microsoft Windows environment and culture for a Portable App to work via automatic detection of a config file within the proximity of the exe file.  Same goes for dll files within the proximity of the exe.  This is a defined standard.  More and more Windows apps behave in this way, because using third-party pseudo-portablizing containment sandbox tools (PortableApps) is very ghey and unnecessary.

olliebean commented:
@a-raccoon : The problem isn't with how the portable mode is enabled, whether by a command line or automatically if the profile folder is detected alongside the exe file. The problem is with the portable mode being unable to correctly handle a user-defined path for where to store the downloaded files without losing track of said files and throwing up errors, and indeed with it moving the user's existing downloaded files out of their user-defined folders and into its profile folder without asking or informing the user. (I think we've established that it does do this, as it has happened to me and to other users on enabling portable mode.)

bacon-cheeseburger commented:
@a-raccoon And a lot of users are **not** Windows users. Rather than trying to shoehorn/force Windows quirks/oddities/etc's, apps should just simply listen to what the user is telling it whether it be by command line, config file, or otherwise. There's no reason an app should push back against user preferences. If the user _explicitly_ instructs the app to behave in a certain way, look for (pro)files in a certain place, etc. then do it. Don't fight the user, just do it.

Another point of consideration for non-Windows/Linux users is many people don't tolerate the idiocy that is distro/filesystem hierarchy. Instead they create more simple & sensible hierarchies that suit their needs and make things like creating backups vastly easier. Generally speaking, puking files all over the place is just bad design.

glassez commented:
>The problem isn't with how the portable mode is enabled

But this PR is about "how the portable mode is enabled". We didn't touch anything else here. So please stop littering this topic with problems that isn't directly related to it.

eli-schwartz commented:
> Another point of consideration for non-Windows/Linux users is many people don't tolerate the idiocy that is distro/filesystem hierarchy. Instead they create more simple & sensible hierarchies that suit their needs and make things like creating backups vastly easier. Generally speaking, puking files all over the place is just bad design.

Why are you backing up program executables? If you're not backing up program executables, then none of this has anything to do with backups.

Complain about portable mode as you will, but simply enabling the mode when faced with a strictly Windows-style layout should really not be affecting Linux users for good or ill. And people doing backups of their profile data should not be using portable mode, as it will puke obtuse, bloated binaries all over their backups, which is just bad design for a backup system.

P.S. Any distro lets you create your own packages (some easier than others), which define whatever you darn well please, no matter how wrong you want to be... but it doesn't matter, qbittorrent is a single executable file, plus icons stored in the icon theme directory and accessed via `QIcon::fromTheme()` to ensure users are permitted to theme the program themselves. So if you really, truly want to relocate /usr/bin/qbittorrent to /opt/lolware/my/path/is/the/best/qbittorrent, you can.

bacon-cheeseburger commented:
@eli-schwartz What are you on about? Nobody said anything about backing up executables but it wouldn't matter if they did. A lot of people *do* in fact backup executables for known-working compatibility reasons. And nobody said anything about relocating "/usr/bin/qbittorrent", if that's even where it's located, to anywhere else. And who cares about creating your own packages using whatever layout suits you if the program doesn't behave properly or in an expected manner according to the given settings?
